File: services/backend/src/data/cliAnythingTemplates/node_skeleton/session.js
'use strict';

class SessionManager {
  static _instance = null;

  constructor() {
    this._state = {};
    this._undoStack = [];
    this._redoStack = [];
  }

  static getCurrent() {
    if (!SessionManager._instance) {
      SessionManager._instance = new SessionManager();
    }
    return SessionManager._instance;
  }

  get state() { return this._state; }

  snapshot() {
    this._undoStack.push(structuredClone(this._state));
    this._redoStack.length = 0;
  }

  update(key, value) {
    this.snapshot();
    this._state[key] = value;
  }

  undo() {
    if (!this._undoStack.length) {
      return { status: 'error', message: 'Nothing to undo' };
    }
    this._redoStack.push(structuredClone(this._state));
    this._state = this._undoStack.pop();
    return { status: 'success', message: 'Undo successful' };
  }

  redo() {
    if (!this._redoStack.length) {
      return { status: 'error', message: 'Nothing to redo' };
    }
    this._undoStack.push(structuredClone(this._state));
    this._state = this._redoStack.pop();
    return { status: 'success', message: 'Redo successful' };
  }

  reset() {
    this._state = {};
    this._undoStack.length = 0;
    this._redoStack.length = 0;
  }
}

module.exports = { SessionManager };

File: services/backend/src/data/cliAnythingTemplates/python_skeleton/session.py
import copy

class SessionManager:
    """Deep-copy undo/redo session manager."""
    _instance = None

    def __init__(self):
        self._state = {}
        self._undo_stack = []
        self._redo_stack = []

    @classmethod
    def get_current(cls):
        if cls._instance is None:
            cls._instance = cls()
        return cls._instance

    @property
    def state(self):
        return self._state

    def snapshot(self):
        """Take a deep copy of current state and push to undo stack."""
        self._undo_stack.append(copy.deepcopy(self._state))
        self._redo_stack.clear()

    def update(self, key, value):
        """Update state with automatic snapshot."""
        self.snapshot()
        self._state[key] = value

    def undo(self):
        if not self._undo_stack:
            return {'status': 'error', 'message': 'Nothing to undo'}
        self._redo_stack.append(copy.deepcopy(self._state))
        self._state = self._undo_stack.pop()
        return {'status': 'success', 'message': 'Undo successful'}

    def redo(self):
        if not self._redo_stack:
            return {'status': 'error', 'message': 'Nothing to redo'}
        self._undo_stack.append(copy.deepcopy(self._state))
        self._state = self._redo_stack.pop()
        return {'status': 'success', 'message': 'Redo successful'}

    def reset(self):
        self._state = {}
        self._undo_stack.clear()
        self._redo_stack.clear()
